proposals/type-classes.md

```diff
@@ -0,0 +1,296 @@
+}
+```
+
+## Composition and chain of evidences
+
+Type class instances and declarations can encode further constrains in their generic args so they can be composed nicely:
+
+```kotlin
+package optionext
+
+extension class OptionMonoid<A>(with Monoid<A>): Monoid<Option<A>> {
+
+  val empty: Option<A> = None
+
+  fun Option.combine(ob: Option<A>): Option<A> =
+    when (this) {
+      is Some<A> -> when (ob) {
+                      is Some<A> -> Some(this.value.combine(b.value)) //works because there is evidence of a Monoid<A>
+                      is None -> ob
+                    }
+      is None -> this
+  }
+    
+}
+```
+
+The above instance declares a `Monoid<Option<A>>` as long as there is a `Monoid<A>` in scope.
+
+```kotlin
+import optionext.OptionMonoid
+import intext.IntMonoid
+
+Option(1).combine(Option(1)) // Option(2)
+Option("a").combine(Option("b")) // does not compile. Found `Monoid<Option<A>>` instance providing `combine` but no `Monoid<String>` instance was in scope
+```
+
+We believe the above proposed encoding fits nicely with Kotlin's philosophy of extensions and will reduce the boilerplate compared to other langs that also support typeclasses such as Scala where this is done via implicits.
+
+## Typeclasses over type constructors
+
+We recommend if this proposal is accepted that a lightweight version of higher kinds support is included to unveil the true power of typeclasses through the extensions mechanisms
+
+A syntax that would allow for higher kinds in these definitions may look like this:
+
+```kotlin
+extension interface FunctionK<F<_>, G<_>> {
+  fun <A> invoke(fa: F<A>): G<A>
+}
+
+extension object Option2List : FunctionK<Option, List> {
+  fun <A> invoke(fa: Option<A>): List<A> =
+    fa.fold({ emptyList() }, { listOf(it) })
+}
+``` 
+
+Here `F<_>` refers to a type constructor meaning a type that has a hole on it such as `Option`, `List`, etc. 
+
+A use of this declaration in a polymorphic function would look like:
+
+```kotlin
+fun <F<_>, A, B> transform(fa: F<A>, f: (A) -> B, with Functor<F>): F<B> = F.map(fa, f)
+
+transform(Option(1), { it + 1 }) // Option(2)
+transform("", { it + "b" }) // Does not compile: `String` is not type constructor with shape F<_>
+transform(listOf(1), { it + 1 }) // does not compile: No `Functor<List>` instance defined in scope.
+```
+
+## Language Changes
+
+- Add `with` to require instances evidences in both function and class/object declarations as demonstrated by previous and below examples:
+```kotlin
+extension class OptionMonoid<A>(with Monoid<A>) : Monoid<Option<A>> //class position using argument "Monoid"
+extension class OptionMonoid<A>(with M: Monoid<A>) : Monoid<Option<A>> //class position using argument "M"
+
+fun <A> add(a: A, b: A, with Monoid<A>): A = a.combine(b) //function position using argument "Monoid"
+fun <A> add(a: A, b: A, with M: Monoid<A>): A = a.combine(b) //function position argument "M"
+```
```

Comment:
What is the difference between named and unnamed variants? Are these two options to choose between?

Reply:
Hi @ilya-g 
the initial idea is to name "Monoid" the `Monoid` argument, now I consider this pretty messy.
"Monoid" is the argument name, the parameter class name or the `Monoid` typeclass.

I consider better treat is as anonymous declaration, so

    fun <A> add(a: A, b: A, with Monoid<A>): A

is same as

    fun <A> add(a: A, b: A, with _: Monoid<A>): A

In such case it is possible to use a type class easy, however if you want to reference to it then you have to declare explicilty its name.